framework/src/modules/chain/rebuilder.js

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright © 2019 Lisk Foundation
+ *
+ * See the LICENSE file at the top-level directory of this distribution
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+'use strict';
+
+const { loadBlocksWithOffset } = require('./blocks');
+
+/**
+ * Rebuild a blockchain
+ * deletes and recalculates all the states from the blocks up to the specified round
+ */
+class Rebuilder {
+	constructor({
+		// components
+		channel,
+		logger,
+		storage,
+		// Unique requirements
+		genesisBlock,
+		// Modules
+		processorModule,
+		blocksModule,
+		interfaceAdapters,
+		// Constants
+		activeDelegates,
+	}) {
+		this.isActive = false;
+		this.isCleaning = false;
+
+		this.channel = channel;
+		this.logger = logger;
+		this.storage = storage;
+		this.genesisBlock = genesisBlock;
+
+		this.processorModule = processorModule;
+		this.blocksModule = blocksModule;
+		this.interfaceAdapters = interfaceAdapters;
+		this.constants = {
+			activeDelegates,
+		};
+	}
+
+	cleanup() {
+		this.isCleaning = true;
+	}
+
+	async rebuild(rebuildUpToRound, loadPerIteration = 1000) {
```

Comment:
After rebuild is finished it doesn't remove blocks with the `height` greater than the `targetHeight`